File: android/src/main/java/com/skjline/rideraid/presenter/module/TeleSensor.java
package com.skjline.rideraid.presenter.module;

import com.skjline.rideraid.ant.AntBikeDevice;
import com.skjline.rideraid.ant.AntDevice;
import com.skjline.rideraid.presenter.TelemetryPresenter;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;

public class TeleSensor implements TelemetryPresenter {
    private AntBikeDevice device;
    private TelemetryPresenter.View view;

    private Disposable holder;

    public TeleSensor(TelemetryPresenter.View view, AntBikeDevice device) {
        this.view = view;
        this.device = device;
    }

    @Override
    public boolean isActive() {
        return holder != null && !holder.isDisposed();
    }

    @Override
    public void start() {
        if (device == null) {
            return;
        }

        initializeBikeEventSubscriber();
    }

    @Override
    public void stop() {
        if (holder == null || !holder.isDisposed()) {
            return;
        }

        holder.dispose();
        holder = null;
    }

    private void initializeBikeEventSubscriber() {
        if (holder != null) {
            return;
        }

        holder = device.getBikeEventObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(bikeEvent -> {
                    switch (bikeEvent.getType()) {
                        case AntBikeDevice.ANT_DEVICE_TYPE_SPEED:
                            view.updateSpeed(bikeEvent.getValue());
                            break;
                        case AntDevice.ANT_DEVICE_TYPE_CADENCE:
                            view.updateCadence(bikeEvent.getValue());
                            break;
                    }
                });
    }
}

File: rider/src/main/java/com/pss9/rideraid/presenter/module/TeleSensor.kt
package com.pss9.rideraid.presenter.module

import com.pss9.rideraid.ant.AntBikeDevice
import com.pss9.rideraid.ant.AntDevice
import com.pss9.rideraid.presenter.TelemetryPresenter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class TeleSensor(private val view: TelemetryPresenter.View, private val device: AntBikeDevice?) : TelemetryPresenter {

    private var holder = CompositeDisposable()

    override val isActive: Boolean
        get() = !holder.isDisposed

    override fun start() {
        if (device == null) {
            return
        }

        initializeBikeEventSubscriber()
    }

    override fun stop() {
        holder.clear()
    }

    private fun initializeBikeEventSubscriber() {
        device?.let {
            holder.add(it.onBikeEvent
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { bikeEvent ->
                    when (bikeEvent.type) {
                        AntDevice.ANT_DEVICE_TYPE_SPEED -> view.updateSpeed(bikeEvent.value)
                        AntDevice.ANT_DEVICE_TYPE_CADENCE -> view.updateCadence(bikeEvent.value)
                    }
                })
        }
    }
}
